skrifa/src/outline/glyf/hint/engine/stack.rs

```diff
@@ -0,0 +1,221 @@
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3199>
+    pub(super) fn op_mindex(&mut self) -> OpResult {
+        self.value_stack.move_index()
+    }
+
+    /// Roll the top three stack elements.
+    ///
+    /// ROLL[] (0x8a)
+    ///
+    /// Pops: a, b, c (top three stack elements)
+    /// Pushes: b, a, c (elements reordered)
+    ///
+    /// Performs a circular shift of the top three objects on the stack with
+    /// the effect being to move the third element to the top of the stack
+    /// and to move the first two elements down one position. ROLL is
+    /// equivalent to MINDEX[] 3.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#roll-the-top-three-stack-elements>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3258>
+    pub(super) fn op_roll(&mut self) -> OpResult {
+        self.value_stack.roll()
+    }
+
+    /// Push data onto the interpreter stack.
+    ///
+    /// NPUSHB[] (0x8a)
+    ///
+    /// Takes n unsigned bytes from the instruction stream, where n is an
+    /// unsigned integer in the range (0..255), and pushes them onto the stack.
+    /// n itself is not pushed onto the stack.
+    ///
+    /// NPUSHW[] (0x41)
+    ///
+    /// Takes n 16-bit signed words from the instruction stream, where n is an
+    /// unsigned integer in the range (0..255), and pushes them onto the stack.
+    /// n itself is not pushed onto the stack.
+    ///
+    /// PUSHB\[abc\] (0xB0 - 0xB7)
+    ///
+    /// Takes the specified number of bytes from the instruction stream and
+    /// pushes them onto the interpreter stack.
+    /// The variables a, b, and c are binary digits representing numbers from
+    /// 000 to 111 (0-7 in binary). Because the actual number of bytes (n) is
+    /// from 1 to 8, 1 is automatically added to the ABC figure to obtain the
+    /// actual number of bytes pushed.
+    ///
+    /// PUSHW\[abc\] (0xB8 - 0xBF)
+    ///
+    /// Takes the specified number of words from the instruction stream and
+    /// pushes them onto the interpreter stack.
+    /// The variables a, b, and c are binary digits representing numbers from
+    /// 000 to 111 (0-7 binary). Because the actual number of bytes (n) is from
+    /// 1 to 8, 1 is automatically added to the abc figure to obtain the actual
+    /// number of bytes pushed.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#pushing-data-onto-the-interpreter-stack>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L3858>
+    pub(super) fn op_push(&mut self, args: &Args) -> OpResult {
```

Comment:
this + args is the only thing so far that raises my eyebrows at all, it seems like a vaguely annoying API? but i'm also not sure where the input comes from, what it looks like in practice etc.

I also of course don't have a better idea; the one alternative that occurs to me is separate `Bytes`/`Words` or `ByteOps`/`WordOps` and then a trait they both implement, but this would only be useful if we could easily construct these types wherever we're getting the ops from, which is probably not so simple.

Basically this is a comment to demonstrate that I am actually reading this PR, and  you can probably ignore it 💁‍♂️

Reply:
This is actually a good call out because I was considering leaving this out of the PR due to the lack of context around the type, but decided to include it to avoid breaking up the `ValueStack` impl.

Hopefully this will make more sense when the remainder of the code lands. Constructing this is actually very simple and is handled by the instruction decoder. The code looks like this:
```rust
        if arg_count > 0 {
            args.bytes = self
                .bytecode
                .get(arg_start..arg_start + arg_count)
                .ok_or(HintErrorKind::UnexpectedEndOfBytecode)?;
            args.is_words = (PUSHW000..=PUSHW111).contains(&opcode)
                || opcode == NPUSHW;
        }
```
where `arg_start` and `arg_count` are derived from the current program counter and a table indexed by opcode. `NPUSHB` and `NPUSHW` read the arg count as the next byte in the instruction stream.

The dispatch code is even simpler, capturing all 18 opcodes in a single match arm and calling out to `op_push` as defined in this PR:

```rust
            PUSHB000..=PUSHW111 | NPUSHB | NPUSHW => {
                self.op_push(&ins.args)?;
            }
```

Thanks for the quick review!